### PR TITLE
Refactor: Add arguments block to main MatNWB api functions

### LINKED ISSUE #606: [Feature]: Improve test coverage to 95%
### What would you like to see added to MatNWB?

See title

### Is your feature request related to a problem?

_No response_

### What solution would you like?

See title

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/neurodatawithoutborders/matnwb/blob/master/.github/CODE_OF_CONDUCT.md)
- [X] Have you ensured this change was not already [requested](https://github.com/NeurodataWithoutBorders/matnwb/issues)?

### PR DESCRIPTION
**Pull request by ehennestad**
Fix #606 

## Motivation
- Use argument blocks for better input validation of user-facing functions
- Improve test coverage

## How to test the behavior?
```
Show here how to reproduce the new behavior (can be a bug fix or a new feature)
```

## Checklist

- [x] Have you ensured the PR description clearly describes the problem and solutions?
- [x] Have you checked to ensure that there aren't other open or previously closed [Pull Requests](https://github.com/neurodatawithoutborders/matnwb/pulls) for the same change?
- [x] If this PR fixes an issue, is the first line of the PR description `fix #XX` where `XX` is the issue number?

### CODE REVIEW

**Comment by bendichter** [comment]
@lawrence-mbf would you be able to review this?

File: +file/cloneNwbFileClass.m
Comment (ehennestad) [inline +file/cloneNwbFileClass.m:1]: - Rename variables for clearer code
- Use fileread instead of fopen, fread, close sequence
- Change name of superclass in call to superclass constructor in the new class def

File: +file/fillExport.m
Comment (ehennestad) [inline +file/fillExport.m:1]: - Remove keyboard (debug statement)
- Negate if statement to remove redundant return statement

File: +file/fillProps.m
Comment (ehennestad) [inline +file/fillProps.m:1]: - Create function `assertValidRefType` to factor out duplicate code
- Reformat `typeStr` with improved capitalisation

**Comment by bendichter** [comment]
@ehennestad looks like there are still conflicts

File: +file/fillValidators.m
Comment (ehennestad) [inline +file/fillValidators.m:1]: - Create local function `getFullClassName` for duplicate code

File: +file/writeNamespace.m
Comment (ehennestad) [inline +file/writeNamespace.m:1]: - Replace exist(name, 'dir') with isfolder(name) for performance and clearer code
- Negate `if` to remove unreachable lines
- Use `onCleanup` to make sure file is closed on errors instead of using try/catch block.

File: +io/+space/+shape/Point.m
Comment (ehennestad) [inline +io/+space/+shape/Point.m:1]: Simplify

File: +io/+spec/+internal/readEmbeddedSpecLocation.m
Comment (ehennestad) [inline +io/+spec/+internal/readEmbeddedSpecLocation.m:1]: Extracted code from nwbRead and added to function `io.spec.internal` namespace to make it more accessible for testing

File: +io/+spec/getEmbeddedSpecLocation.m
Comment (ehennestad) [inline +io/+spec/getEmbeddedSpecLocation.m:1]: Extracted code from nwbRead and added to function `io.spec.internal` namespace to make it more accessible for testing

File: +io/+spec/readEmbeddedSpecifications.m
Comment (ehennestad) [inline +io/+spec/readEmbeddedSpecifications.m:1]: Extracted code from nwbRead and added to function `io.spec` namespace to make it more accessible for testing

File: +io/+spec/writeEmbeddedSpecifications.m
Comment (ehennestad) [inline +io/+spec/writeEmbeddedSpecifications.m:1]: Extracted function from `NwbFile/export` to collect all code related to reading and writing embedded specifications in one place

File: +io/parseGroup.m
Comment (ehennestad) [inline +io/parseGroup.m:1]: Moved rehash statement to `file.cloneNwbFileClass`

File: +io/writeAttribute.m
Comment (ehennestad) [inline +io/writeAttribute.m:1]: Use `onCleanup` to close space and object instead of doing it "conditionally"

File: +matnwb/+common/findLatestSchemaVersion.m
Comment (ehennestad) [inline +matnwb/+common/findLatestSchemaVersion.m:1]: Add utility function for finding the latest available nwb schema version

File: +matnwb/+common/mustBeNwbFile.m
Comment (ehennestad) [inline +matnwb/+common/mustBeNwbFile.m:1]: Add validator function to assert that a path is an NWB filepath

File: +matnwb/+common/mustBeValidSchemaVersion.m
Comment (ehennestad) [inline +matnwb/+common/mustBeValidSchemaVersion.m:1]: Add validator function to assert that a `versionNumber` value is a properly formatted version number string and that the version number is available among the available nwb schemas

File: +schemes/Namespace.m
Comment (ehennestad) [inline +schemes/Namespace.m:1]: - Specify default values in property block instead of constructor
- Use assert instead of if to improve coverage

File: +spec/generate.m
Comment (ehennestad) [inline +spec/generate.m:1]: - Also allow `schemaSource` being a string 
- use fileread instead of fopen, fwrite, fclose sequence

File: +spec/loadCache.m
Comment (ehennestad) [inline +spec/loadCache.m:1]: - Add named arguments and arguments block to simplify argument parsing
- Negate `isempty(fileList)` block to remove potentially unreachable code

File: +tests/+system/NWBFileIOTest.m
Comment (ehennestad) [inline +tests/+system/NWBFileIOTest.m:1]: Add unit tests  to increase coverage for `nwbRead`, `nwbExport`

File: +tests/+unit/+file/CloneNwbTest.m
Comment (ehennestad) [inline +tests/+unit/+file/CloneNwbTest.m:1]: Add unites to test the `file.cloneNwbFileClass` function

File: +tests/+unit/dataPipeTest.m [hunk 1]
```diff
@@ -131,6 +142,9 @@ function testExternalFilters(testCase)
     import types.untyped.datapipe.properties.DynamicFilter;
     import types.untyped.datapipe.properties.Shuffle;
     
+    % TODO: Why is Filter.LZ4 not part of the exported Pipe, i.e when the
```
Comment (ehennestad) [inline +tests/+unit/dataPipeTest.m:145]: I need help on this

File: +types/+hdmf_common/DynamicTableRegion.m
Comment (ehennestad) [inline +types/+hdmf_common/DynamicTableRegion.m:1]: Result of updating class generator to fix word capitalisation

File: +types/+untyped/+datapipe/+properties/DynamicFilter.m
Comment (ehennestad) [inline +types/+untyped/+datapipe/+properties/DynamicFilter.m:1]: Added arguments block

File: +types/+untyped/+datapipe/BoundPipe.m
Comment (ehennestad) [inline +types/+untyped/+datapipe/BoundPipe.m:1]: Move default out put to beginning as it was potentially unreachable for unit tests

File: +types/+untyped/Set.m
Comment (ehennestad) [inline +types/+untyped/Set.m:1]: Removed unused method

File: +types/+util/+dynamictable/checkConfig.m
Comment (ehennestad) [inline +types/+util/+dynamictable/checkConfig.m:1]: Added named arguments and argument block

File: +types/+util/+dynamictable/clear.m
Comment (ehennestad) [inline +types/+util/+dynamictable/clear.m:1]: Fixed to work for older schema versions

File: NwbFile.m
Comment (ehennestad) [inline NwbFile.m:1]: - Add arguments block to constructor and export method
- Added mode for overwriting file on export
- Use assertion instead of if block with error on export to improve coverage
- removed `embedSpecifications` method. Was moved to `io.spec.writeEmbeddedSpecifications`

File: functionSignatures.json
Comment (ehennestad) [inline functionSignatures.json:1]: moved file into resources folder

File: generateCore.m
Comment (ehennestad) [inline generateCore.m:1]: - Add named arguments and arguments block
- Pass `namespaceFiles` cell array to generateExtension instead of potentially calling it twice

File: generateExtension.m
Comment (ehennestad) [inline generateExtension.m:1]: - Add named arguments and arguments block
- Added local validator function `mustBeYamlFle` to check that namespace spec files exist
- simplified code in loop for generating class files

File: nwbExport.m
Comment (ehennestad) [inline nwbExport.m:1]: - Rename input arguments to be more descriptive
- Added argument blocks for simpler input handling / validation

File: resources/functionSignatures.json
Comment (ehennestad) [inline resources/functionSignatures.json:1]: Moved to resources folder
Changed to be up-to-date

File: +io/writeAttribute.m [hunk 2]
```diff
@@ -3,29 +3,32 @@ function writeAttribute(fid, fullpath, data, varargin)
...
     H5A.write(id, tid, data);
 end
 H5A.close(id);
-H5S.close(sid);
-H5O.close(oid);
+
+function closeSpaceAndObject(spaceId, objectId)
+    H5S.close(spaceId);
+    H5O.close(objectId);
+end
```
Comment (bendichter) [inline +io/writeAttribute.m:33]: This feels like it could have some side effects. Are you able to separate this change into a different PR?

**Review by bendichter — APPROVED** [review]
